### PR TITLE
[fsdp, trainer] feat: Run vLLM-based Generative Reward Model with Multi-Workgroup Support

### PR DESCRIPTION
**Pull request by imboaba**
Solve problem #1193 by multiple vllm instance available.

- The core enhancement is **decoupling the resource pool from the WorkerGroup**, allow roles to be assigned to separate WorkerGroups while sharing a single resource pool. Consequently, multiple vLLM rollout engines can be launched without partitioning GPU resources.
- This feature allows users to load a extra vllm instance to compute rewards. By loading custom `reward_preprocess` and `reward_postprocess` functions, users can define custom reward prompts or even integrate a reward function directly with the reward model.
- It is easy to support other rollout engine in reward model, and can be used as a foundation to develop new features like [critic-rl](https://github.com/HKUNLP/critic-rl/tree/main)

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: 
#1953 
#2748 
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```bash
# Add code snippet or script demonstrating how to use this
bash example/reward_model/run_qwen3-1.7b_rm.sh
```

### Design & Code Changes

> 1. use `group_id` to specify WorkerGroup, one group for one rollout engine

```python
# main_ppo_rm.py

@ray.remote(num_cpus=1)  # please make sure main_task is not scheduled on head
class TaskRunner:
    def run(self, config):
        ... 
        global_pool_id = "global_pool"
        global_group_id = "global_group"
        rm_group_id = "rm_group"

        resource_pool_spec = {
            global_pool_id: [config.trainer.n_gpus_per_node] * config.trainer.nnodes,
        }

        # use group_id to map the roles to another group
        role_config_list: List[RoleLayoutConfigItem] = [
            RoleLayoutConfigItem(
                Role.ActorRollout, ray.remote(actor_rollout_cls),
                global_pool_id, global_group_id,
            ),
            RoleLayoutConfigItem(
                Role.Critic, ray.remote(CriticWorker),
                global_pool_id, global_group_id,
            ),
        ]

        if config.reward_model.enable:
            from verl.workers.fsdp_workers import RewardModelWorker
            role_config_list.append(RoleLayoutConfigItem(
                Role.RewardModel, ray.remote(RewardModelWorker),
                global_pool_id, rm_group_id,
            ))
        
        ...
        
        # use SharedPoolRayPPOTrainer for multi-workgroup
        trainer = SharedPoolRayPPOTrainer(
            config=config,
            tokenizer=tokenizer,
            processor=processor,
            role_layout_config=role_layout_config,
            resource_pool_manager=resource_pool_manager,
            ray_worker_group_cls=ray_worker_group_cls,
            reward_fn=reward_fn,
            val_reward_fn=val_reward_fn,
            train_dataset=train_dataset,
            val_dataset=val_dataset,
            collate_fn=collate_fn,
            train_sampler=train_sampler,
            device_name=config.trainer.device,
        )
        trainer.init_workers()
        trainer.fit()
```

> 2. separate WorkerGroup by group_id

```python

class ColocateResourcePoolManager(ResourcePoolManager):

    def create_resource_pool(self, pool_map: Dict[str, Set]):
        for resource_pool_name, process_on_nodes in self.resource_pool_spec.items():
            assert resource_pool_name in pool_map.keys()
            max_colocate_count = len(pool_map[resource_pool_name])
            # set max_colocate_count
            resource_pool = RayResourcePool(process_on_nodes=process_on_nodes, use_gpu=True, max_colocate_count=max_colocate_count, name_prefix=resource_pool_name)
            self.resource_pool_dict[resource_pool_name] = resource_pool

        self._check_resource_available()

class SharedPoolRayPPOTrainer(RayPPOTrainer):
    ...
    def init_workers(self):
        # Create the resource pool to be shared among different worker groups.
        pool_2_group = {pool: set() for pool in self.role_layout_config.resource_pools}
        for item in self.role_layout_config.role_map.values():
            pool_2_group[item.resource_pool_name].add(item.group_name)
        self.resource_pool_manager.create_resource_pool(pool_map=pool_2_group)

        ...
        for work_group_id in self.role_layout_config.groups:
            roles: List[Role] = [item.role for item in self.role_layout_config.role_map.values() 
                     if item.group_name == work_group_id]

            group_class_dict = ......
            
            worker_dict_cls = create_colocated_worker_cls(class_dict=group_class_dict)

```

> 3. initialize the rollout engine in reward model worker

```python
# fsdp_worker.py
class RewardModelWorker(Worker, DistProfilerExtension):
    @register(dispatch_mode=Dispatch.ONE_TO_ALL)
    def init_model(self):
        # This is used to import external_lib into the huggingface systems
        import_external_libs(self.config.model.get("external_lib", None))
        self.reward_module, self.model_config = self._build_model(config=self.config)
        if self.use_rollout:
            self.rollout_engine, self.rollout_sharding_manager = self._build_rollout(
                trust_remote_code=self.config.model.get("trust_remote_code", False)
            )
```

> 4. use `reward_preprocess` and `reward_postprocess` to process the reward data

```python
# fsdp_worker.py
class RewardModelWorker(Worker, DistProfilerExtension):
    @register(dispatch_mode=Dispatch.DP_COMPUTE_PROTO)
    @DistProfiler.annotate(color="brown")
    def compute_rm_score(self, data: DataProto):
        ...
        if self.GenRM_mode:
            questions, answers, ground_truths = self._extract_inputs_for_reward(data)
            chats, scores_tensor = self.reward_preprocess_fn(
                questions, answers, ground_truths
            )
            rm_data = self._chats_to_ids(chats, data.batch['attention_mask'].shape[-1])
            if scores_tensor is not None:
                scores_tensor = scores_tensor.to(rm_data.batch.device)
                rm_data.batch['extra_scores'] = scores_tensor

        ...
            for micro_batch in micro_batches:
                if self.GenRM_mode:
                    output_proto = self._forward_rollout_micro_batch(micro_batch)
                    generated_texts = self.tokenizer.batch_decode(
                        output_proto.batch["responses"],
                        skip_special_tokens=True
                    )
                    # input: List[str], output: tensor
                    rm_score = self.reward_postprocess_fn(
                        generated_texts,
                        micro_batch.get("extra_scores", None)
                    )
        ...
```

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
@yyDing1 could u help review?

**Comment by yyDing1** [comment]
> @yyDing1 could u help review?

Yes, I will review this feat.

**Comment by yyDing1** [comment]
@imboaba Thank you for your great contribution. Decoupling the resource pool from the WorkerGroup is a valuable effort. I have the following suggestions:

- Can you simplify "reward_process" by only passing necessary parameters (e.g., question, answer, ground_truth, genrm_responses) and returning the chats_messages or final reward tensor (like `customized_reward_function`)? This might be more friendly for user-customized scenarios. The process of DataProto can be placed in the inner workers part.

- The parameter name "use_rollout" might cause misunderstanding, as it could be interpreted as directly using the rollout model for reward calculation. It may be helpful to add some clarification or rename this parameter.

- It’s beneficial to add some CI tests for this new feature.

**Comment by vermouth1992** [comment]
I guess it's better to wait for the ModelEngine refactor and RolloutEngine refactor to continue this functionality. To implement GenRM within hybridengine, we will need to have a separate RolloutWorker so that the code is clean and manageable.

**Comment by imboaba** [comment]
> @imboaba Thank you for your great contribution. Decoupling the resource pool from the WorkerGroup is a valuable effort. I have the following suggestions:
> 
> * Can you simplify "reward_process" by only passing necessary parameters (e.g., question, answer, ground_truth, genrm_responses) and returning the chats_messages or final reward tensor (like `customized_reward_function`)? This might be more friendly for user-customized scenarios. The process of DataProto can be placed in the inner workers part.
> * The parameter name "use_rollout" might cause misunderstanding, as it could be interpreted as directly using the rollout model for reward calculation. It may be helpful to add some clarification or rename this parameter.
> * It’s beneficial to add some CI tests for this new feature.

Hi @yyDing1, thank you for your suggestions. Following your advice, I have refactored and simplified parts of the reward_process. Additionally, CI tests for this new feature have been added and have passed locally.

**Comment by imboaba** [comment]
> I guess it's better to wait for the ModelEngine refactor and RolloutEngine refactor to continue this functionality. To implement GenRM within hybridengine, we will need to have a separate RolloutWorker so that the code is clean and manageable.

You're right. In fact, the current rollout logic is pretty much copied from `actor_rollout_ref`. A dedicated RolloutWorker would make things a lot simpler.
